pymc3/distributions/multivariate.py

```diff
@@ -93,35 +61,64 @@ class MvNormal(Continuous):
 
     """
 
-    def __init__(self, mu, cov=None, tau=None, gpu_compat=False, *args, **kwargs):
+    def __init__(self, mu, cov=None, tau=None, chol=None, gpu_compat=False,
+                 *args, **kwargs):
         super(MvNormal, self).__init__(*args, **kwargs)
+        if len([i for i in [tau, cov, chol] if i is not None]) != 1:
+            raise ValueError('Incompatible parameterization. Specify exactly '
+                             'one of tau, cov, or chol to specify '
+                             'distribution.')
         self.mean = self.median = self.mode = self.mu = mu = tt.as_tensor_variable(mu)
-        tau, cov = get_tau_cov(mu, tau=tau, cov=cov)
-        self.tau = tt.as_tensor_variable(tau)
-        self.cov = tt.as_tensor_variable(cov)
+        self.solve = tt.slinalg.Solve(A_structure="lower_triangular", lower=True)
+
+        # Use the Cholesky logp unless tau is given
+        self.use_chol_logp = tau is None
+        if cov is not None:
+            self.chol = tt.slinalg.cholesky(tt.as_tensor_variable(cov))
+        elif tau is not None:
+            self.tau = tt.as_tensor_variable(tau)
+            self.chol = tt.slinalg.cholesky(tt.nlinalg.matrix_inverse(tau))
+        else:
+            self.chol = tt.as_tensor_variable(chol)
+
         self.gpu_compat = gpu_compat
         if gpu_compat is False and theano.config.device == 'gpu':
             warnings.warn("The function used is not GPU compatible. Please check the gpu_compat flag")
 
     def random(self, point=None, size=None):
-        mu, cov = draw_values([self.mu, self.cov], point=point)
-
-        def _random(mean, cov, size=None):
-            return stats.multivariate_normal.rvs(
-                mean, cov, None if size == mean.shape else size)
-
-        samples = generate_samples(_random,
-                                   mean=mu, cov=cov,
-                                   dist_shape=self.shape,
-                                   broadcast_shape=mu.shape,
-                                   size=size)
-        return samples
+        mu, chol = draw_values([self.mu, self.chol], point=point)
+        if size is None or size == mu.shape:
+            size = mu.shape
+        elif isinstance(size, collections.Iterable):
```

Comment:
@aseyboldt current behavior is to sample params in advance and pass them to variable sampler. I write #1928 with intuition that there should be a static sampler along with context dependent. Then it will be possible to add theano Op to any random sampler moreover I sample context automatically there.